lib/puppet/configurer.rb

```diff
@@ -55,13 +56,20 @@ def initialize(factory = Puppet::Configurer::DownloaderFactory.new)
   # Get the remote catalog, yo.  Returns nil if no catalog can be found.
   def retrieve_catalog(query_options)
     query_options ||= {}
-    # First try it with no cache, then with the cache.
-    unless (Puppet[:use_cached_catalog] and result = retrieve_catalog_from_cache(query_options)) or result = retrieve_new_catalog(query_options)
-      if ! Puppet[:usecacheonfailure]
-        Puppet.warning "Not using cache on failed catalog"
-        return nil
+    if (Puppet[:use_cached_catalog] && result = retrieve_catalog_from_cache(query_options))
+      @cached_catalog_status = 'used because use_cached_catalog was set'
+    else
+      result = retrieve_new_catalog(query_options)      
+
+      if !result 
+        if ! Puppet[:usecacheonfailure]
```

Comment:
Style thing: I'd like to be consistent about whether we have a space after the `!` or not (and the above two lines differ). I'm sure the puppet codebase overall is *super* inconsistent, but at least within a PR or file there's a charm in being consistent.

Fwiw, rubocop prefers no-space after the `!`: https://github.com/bbatsov/rubocop/blob/master/config/enabled.yml#L659-L662